// osjs/osjs.js
/*!
 * OS.js - JavaScript Cloud/Web Desktop Platform
 *
 * Copyright (c) 2011-2016, Anders Evenrud <andersevenrud@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @author  Anders Evenrud <andersevenrud@gmail.com>
 * @licence Simplified BSD License
 */
(function(_path) {
  'use strict';

  var api = require('./api.js');
  var vfs = require('./vfs.js');

  /////////////////////////////////////////////////////////////////////////////
  // GLOBALS
  /////////////////////////////////////////////////////////////////////////////

  var namespace = {};
  var config = {};
  var handler = {
    register: function(CONFIG, API, HANDLER) {
      API.login = function(data) {
        return data;
      };
      API.logout = function() {
        return true;
      };
    },
    checkPrivilege: function() {
      return true;
    }
  };

  /////////////////////////////////////////////////////////////////////////////
  // METHODS
  /////////////////////////////////////////////////////////////////////////////

  /**
   * API Request proxy
   */
  function request(isVfs, method, args, callback, req, response) {
    response = response || {
      statusCode: 200
    };

    req = req || {
      cookies: {
        get: function(k) {
          if ( k === 'username' ) {
            return 'demo';
          } else if ( k === 'groups' ) {
            return ['admin'];
          }
          return null;
        },
        set: function() {
        }
      }
    };

    if ( isVfs ) {
      if ( vfs[method] && ((['getMine', 'getRealPath']).indexOf(method) < 0) ) {
        vfs[method](args, req, function(json) {
          if ( !json ) {
            json = { error: 'No data from response' };
          }

          callback(json.error, json.result);
        }, config);
      } else {
        throw 'Invalid VFS method: ' + method;
      }

      return;
    }

    if ( namespace[method] ) {
      namespace[method](args, callback, req, response, config);
    } else {
      throw 'Invalid method: ' + method;
    }
  }

  /////////////////////////////////////////////////////////////////////////////
  // EXPORTS
  /////////////////////////////////////////////////////////////////////////////

  /**
   * Initializes OS.js APIs and configurations
   *
   * @param   Object    setup       Configuration
   *
   * @option  setup     int       port        Listening port (default=null/auto)
   * @option  setup     String    dist        Build name (default=dist)
   * @option  setup     String    dirname     Server running dir (ex: /osjs/src/server/node)
   * @option  setup     String    root        Installation root directory (ex: /osjs)
   * @option  setup     boolean   nw          NW build (default=false)
   * @option  setup     boolean   logging     Enable logging (default=true)
   * @option  setup     Mixed     settings    (Optional) Auto-detected. Path to JSON file or Object
   * @option  setup     String    repodir     (Optional) Auto-detected. Package repository root directory (ex: /osjs/src/packages)
   * @option  setup     String    distdir     (Optional) Auto-detected. Build root directory (ex: /osjs/dist)
   *
   * @return  Object                          Returns an object with `api`, `vfs`, `request`, `handler` and `config` helpers
   *
   * @api     osjs.init
   */
  module.exports.init = function(setup) {
    setup.dist     = setup.dist     || 'dist';
    setup.settings = setup.settings || _path.join(_path.dirname(setup.dirname), 'settings.json'),
    setup.repodir  = setup.repodir  || _path.join(setup.root, 'src', 'packages');
    setup.distdir  = setup.distdir  || _path.join(setup.root, setup.dist);
    setup.logging  = typeof setup.logging === 'undefined' || setup.logging === true;

    if ( setup.nw ) {
      setup.repodir = _path.join(setup.root, 'packages');
      setup.distdir = setup.root;
    }

    config = require('./config.js').init(setup);

    // Register internal API methods
    var privilegeMap = {
      application: 'application',
      curl: 'curl',
      fs: 'vfs'
    };

    Object.keys(api).forEach(function(fn) {
      namespace[fn] = function(args, callback, request, response) {
        if ( privilegeMap[fn] ) {
          handler.checkPrivilege(request, response, privilegeMap[fn]);
        }
        api[fn].apply(api[fn], arguments);
      };
    });

    // Register handler API methods
    if ( !setup.nw ) {
      var hs = _path.join(setup.dirname, 'handlers', config.handler, 'handler.js');
      var ihs = require(hs);
      if ( ihs.checkPrivilege ) {
        handler.checkPrivilege = ihs.checkPrivilege;
      }
      ihs.register(config, namespace, handler);
    }

    if ( !handler.checkPrivilege ) {
      handler.checkPrivilege = function(request, response, privilege) {
        if ( setup.nw ) {
          return true;
        }

        if ( typeof privilege !== 'boolean' ) {
          if ( !privilege ) {
            privilege = [];
          }
          if ( !(privilege instanceof Array) && privilege ) {
            privilege = [privilege];
          }
        }

        function check() {
          var groups = [];
          try {
            groups = JSON.parse(request.cookies.get('groups'));
          } catch ( e ) {
            groups = [];
          }

          if ( groups.indexOf('admin') < 0 ) {
            var allowed = true;
            privilege.forEach(function(p) {
              if ( groups.indexOf(p) < 0 ) {
                allowed = false;
              }
              return allowed;
            });
            return allowed;
          }

          return true;
        }

        var uname = request.cookies.get('username');
        if ( !uname ) {
          throw 'You have no OS.js Session, please log in!';
        }
        if ( privilege.length && !check() ) {
          throw 'You are not allowed to use this API function!';
        }

        return true;
      };
    }

    // Register package extension API methods
    if ( config.extensions ) {
      var exts = config.extensions;
      exts.forEach(function(f) {
        if ( f.match(/\.js$/) ) {
          require(config.rootdir + f).register(config, namespace, handler);
        }
      });
    }

    return {
      api: namespace,
      vfs: vfs,
      request: request,
      handler: handler,
      config: config
    };
  };

})(require('path'));
